packages/contrast-colors/index.js

```diff
@@ -13,7 +13,480 @@ governing permissions and limitations under the License.
+    // Then we can run the code for Color, but we've added in more below.
+    Color.prototype._generateColorScale.call(this);
+
+    // create massive scale
+    let backgroundColorScale = createScale({swatches: 1000, colorKeys: this._colorKeys, colorspace: this._colorspace, shift: 1, smooth: this._smooth});
+
+    // Inject original keycolors to ensure they are present in the background options
+    backgroundColorScale.push(this.colorKeys);
+
+    let colorObj = backgroundColorScale
+      // Convert to HSLuv and keep track of original indices
+      .map((c, i) => { return { value: Math.round(cArray(c)[2]), index: i } });
+
+    let bgColorArrayFiltered = removeDuplicates(colorObj, "value")
+      .map(data => backgroundColorScale[data.index]);
+
+    // Manually cap the background array at 100 colors, then add white back to the end
+    // since it sometimes gets removed.
+    bgColorArrayFiltered.length = 100;
+    bgColorArrayFiltered.push('#ffffff');
+
+    this._backgroundColorScale = bgColorArrayFiltered.map(color => {
+      return convertColorValue(color, this._output);
+    });
+
+    return this._backgroundColorScale;
+  }
+}
+
+
+/** 
+ * Utility functions
+ */
+
+function multiplyRatios(ratio, multiplier) {
+  let r;
+  // Normalize contrast ratios before multiplying by this._contrast
+  // by making 1 = 0. This ensures consistent application of increase/decrease
+  // in contrast ratios. Then add 1 back to number for contextual ratio value.
+  if(ratio > 1) {
+    r = ((ratio-1) * multiplier) + 1;
+  }
+  else if(ratio < -1) {
+    r = ((ratio+1) * multiplier) - 1;
+  }
+  else {
+    r = 1;
+  }
+
+  return Number(r.toFixed(2));
+}
+
+function createScale({
+  swatches,
+  colorKeys,
+  colorspace = 'LAB',
+  shift = 1,
+  fullScale = true,
+  smooth = false
+} = {}) {
+  const space = colorSpaces[colorspace];
+  if (!space) {
+    throw new Error(`Colorspace “${colorspace}” not supported`);
+  }
+  if (!colorKeys) {
+    throw new Error(`Colorkeys missing: returned “${colorKeys}”`);
+  }
+  
+  let domains = colorKeys
+    .map(key => swatches - swatches * (d3.hsluv(key).v / 100))
+    .sort((a, b) => a - b)
+    .concat(swatches);
+
+  domains.unshift(0);
+
+  // Test logarithmic domain (for non-contrast-based scales)
+  let sqrtDomains = d3.scalePow()
+    .exponent(shift)
+    .domain([1, swatches])
+    .range([1, swatches]);
+
+  sqrtDomains = domains.map((d) => {
+    if (sqrtDomains(d) < 0) {
+      return 0;
+    }
+    return sqrtDomains(d);
+  });
+
+  // Transform square root in order to smooth gradient
+  domains = sqrtDomains;
+
+  let sortedColor = colorKeys
+    // Convert to HSLuv and keep track of original indices
+    .map((c, i) => { return { colorKeys: cArray(c), index: i } })
+    // Sort by lightness
+    .sort((c1, c2) => c2.colorKeys[2] - c1.colorKeys[2])
+    // Retrieve original RGB color
+    .map(data => colorKeys[data.index]);
```

Comment:
It would be interesting to compare performance with a single sort call. You'd be converting to HSLuv many more times, but you would remove two `O(n)` iterations

```js
let sortedColor = colorKeys
  .sort((c1, c2) => {
    return cArray(c2) - cArray(c1);
  });
```

This approach will certainly use less memory and have way less objects on the heap, but it could be slower depending on how fast `cArray` is.